src/vm_cpu.rs

```diff
@@ -96,6 +99,270 @@ pub struct VmCpuTrapState {
+            let result = self.mask & (1 << self.index);
+            self.mask &= !(1 << self.index);
+            self.index += 1;
+            Some(result)
+        } else {
+            None
+        }
+    }
+}
+
+impl VmPmuState {
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to configure counters.
+    pub fn update_configured_counter(
+        &mut self,
+        counter_index: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) {
+        let counter_index = counter_index as usize;
+        let new_state = CounterState {
+            config_flags,
+            event_type,
+            event_data,
+            value: 0,
+        };
+        let mut updated = false;
+        if let Some(state) = self.counter_state.iter_mut().nth(counter_index) {
+            updated = match state {
+                // If skip_match is set, counter must already be configured
+                Configured(mut c) | Started(mut c) if config_flags.get_skip_match() => {
+                    if config_flags.get_clear_value() {
+                        c.value = 0;
+                    }
+                    if config_flags.get_auto_start() {
+                        *state = Started(c);
+                    }
+                    true
+                }
+                NotConfigured if !config_flags.get_skip_match() => {
+                    *state = if config_flags.get_auto_start() {
+                        Started(new_state)
+                    } else {
+                        Configured(new_state)
+                    };
+                    true
+                }
+                _ => false,
+            };
+        }
+        if !updated {
+            println!(
+                "Warning: Configure counter update failed for counter index {counter_index}, {}",
+                config_flags.raw()
+            );
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to start counters.
+    pub fn update_started_counters(
+        &mut self,
+        counter_index: u64,
+        counter_mask: u64,
+        start_flags: PmuCounterStartFlags,
+        initial_value: u64,
+    ) {
+        let bitmask_iter = BitmaskIter::new(counter_mask);
+        let mut updated = false;
+        self.counter_state
+            .iter_mut()
+            .skip(counter_index as usize)
+            .zip(bitmask_iter)
+            .for_each(|state_and_mask| {
+                let state = state_and_mask.0;
+                let selected = state_and_mask.1;
+                if let Configured(c) = state && selected == 1 {
+                    if start_flags.get_init_value() {
+                        c.value = initial_value;
+                    }
+                    *state = Started(*c);
+                    updated = true;
+                }
+            });
```

Comment:
If you take my suggestion above re: the iterator, a more concise way to write this would be (IMO):

```
for index in BitmaskIter::new(counter_index, counter_mask) {
    if let Configured(c) = &mut self.counter_state[index] {
        ...
    }
}
```

You could also have `BitmaskIter::new()` do the bounds checking for `counter_mask` and `counter_index` so that you can assume indices are valid inside your loop.